service/controller/app/v1/resource/chartoperator/resource.go

```diff
@@ -154,30 +134,100 @@ func (r Resource) installChartOperator(ctx context.Context, cr v1alpha1.App) err
 		}()
 	}
 
-	var chartOperatorValues []byte
 	{
-		values, err := r.values.MergeAll(ctx, *chartOperatorAppCR, *appCatalogCR)
+		err = cc.HelmClient.InstallReleaseFromTarball(ctx, tarballPath, namespace, helm.ReleaseName(release), helm.ValueOverrides(chartOperatorValues))
 		if err != nil {
 			return microerror.Mask(err)
 		}
+	}
 
-		chartOperatorValues, err = json.Marshal(values)
+	{
+		// We wait for the chart-operator deployment to be ready so the
+		// chart CRD is installed. This allows the chart
+		// resource to create CRs in the same reconcilation loop.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for ready chart-operator deployment")
+
+		o := func() error {
+			err := r.checkDeploymentReady(ctx, cc.K8sClient)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+
+		// Wait for chart-operator to be deployed. If it takes longer than
+		// the timeout the chartconfig CRs will be created during the next
+		// reconciliation loop.
+		b := backoff.NewConstant(20*time.Second, 10*time.Second)
+		n := func(err error, delay time.Duration) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("%#q deployment is not ready retrying in %s", release, delay), "stack", fmt.Sprintf("%#v", err))
+		}
+
+		err = backoff.RetryNotify(o, b, n)
 		if err != nil {
 			return microerror.Mask(err)
 		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "chart-operator deployment is ready")
+	}
+
+	return nil
+}
+
+func (r Resource) updateChartOperator(ctx context.Context, cr v1alpha1.App) error {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	chartOperatorAppCR, err := r.getChartOperatorAppCR(ctx, cr.Namespace)
+	if err != nil {
+		return microerror.Mask(err)
 	}
 
+	appCatalogCR, err := r.getAppCatalogCR(ctx, chartOperatorAppCR)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	chartOperatorValues, err := r.mergeChartOperatorValues(ctx, chartOperatorAppCR, appCatalogCR)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// check app CR for chart-operator and fetching app-catalog name and version.
+	var tarballURL string
 	{
-		err = cc.HelmClient.InstallReleaseFromTarball(ctx, tarballPath, namespace, helm.ReleaseName(release), helm.ValueOverrides(chartOperatorValues))
+		tarballURL, err = tarball.NewURL(key.AppCatalogStorageURL(*appCatalogCR), release, key.Version(*chartOperatorAppCR))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var tarballPath string
+	{
+		tarballPath, err = cc.HelmClient.PullChartTarball(ctx, tarballURL)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		defer func() {
+			err := r.fileSystem.Remove(tarballPath)
+			if err != nil {
+				r.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("deletion of %#q failed", tarballPath), "stack", fmt.Sprintf("%#v", err))
+			}
+		}()
+	}
```

Comment:
Pulling the chart tarball is also repeated for install and update. I think it can also be extracted to a `r.pullChartOperatorTarball` method.

Reply:
@rossf7 My only concern is that we are using `defer` to delete the tarball file after use. Even If we would move it as another function, that means we are repeating `defer` into two places which I don't think ideal. WDYT?

Reply:
Ack. Good point about the defer.